mlops/orchestration/airflow/dags/bclaws_image_scraper_dag.py

```diff
@@ -41,166 +49,160 @@
+class ImageCollector:
+    def __init__(self):
+        self.image_map: Dict[str, Set[str]] = defaultdict(set)
+        
+    def add_image(self, xml_path: str, image_url: str):
+        """Add image URL to the collection, organized by XML source"""
+        self.image_map[xml_path].add(image_url)
+    
+    def get_all_images(self) -> Dict[str, Set[str]]:
+        return self.image_map
+
+class ImageFilter:
+    @staticmethod
+    def is_low_quality(filename: str, all_urls: Set[str]) -> bool:
+        """Determine if an image is low quality based on defined rules"""
+        if any(filename.lower().startswith(prefix) for prefix in REMOVE_PREFIXES):
+            return True
+            
+        if "_sm_" in filename or "_sm-" in filename:
+            high_quality = filename.replace("_sm_", "_lg_").replace("_sm-", "_lg-")
+            return any(high_quality in url for url in all_urls)
+            
+        if re.search(r"_\.[^.]+$", filename):
+            base_name = re.sub(r"_(\.[^.]+)$", r"\1", filename)
+            return any(base_name in url for url in all_urls)
+            
+        return False
+
+# ================================
+# Core Functions
+# ================================
+
+def collect_image_urls() -> ImageCollector:
+    """Traverse all XML files and collect image URLs"""
+    collector = ImageCollector()
+    xml_folder = os.path.join(BASE_PATH, XML_DIR)
+    
+    for root, _, files in os.walk(xml_folder):
+        for file in files:
+            if file.endswith('.xml'):
+                xml_path = os.path.join(root, file)
+                try:
+                    tree = etree.parse(xml_path)
+                    for element in tree.iter():
+                        href = element.get('href')
+                        if href and any(href.lower().endswith(ext) for ext in IMAGE_EXTENSIONS):
+                            if "/statreg/" in href:
+                                image_name = href.split('/statreg/')[-1]
+                                image_url = BASE_URL + image_name
+                                collector.add_image(xml_path, image_url)
+                except Exception as e:
+                    print(f"Error processing {xml_path}: {e}")
+    
+    return collector
+
+def filter_images(collector: ImageCollector) -> Dict[str, List[str]]:
+    """Filter out low-quality images from the collection"""
+    filtered_images = {}
+    image_filter = ImageFilter()
+    
+    for xml_path, urls in collector.get_all_images().items():
+        all_urls = set(urls)
+        filtered_urls = [
+            url for url in urls
+            if not image_filter.is_low_quality(url.split('/')[-1], all_urls)
+        ]
+        filtered_images[xml_path] = filtered_urls
+    
+    return filtered_images
 
 @retry(**RETRY_CONFIG)
-def download_image(url, save_path):
-    """Download an image and save it locally."""
+def download_image(url: str, save_path: str):
+    """Download an image with rate limiting"""
```

Comment:
interesting.  Why do we need rate limiting?

Reply:
so to not overwhelm the api and get prematurely stopped from downloading!